File: 2904-shortest-and-lexicographically-smallest-beautiful-string/2904-shortest-and-lexicographically-smallest-beautiful-string.js
/**
 * @param {string} s
 * @param {number} k
 * @return {string}
 */
var shortestBeautifulSubstring = function (s, k) {
    let ans = "";
    let left = 0;
    let ones = 0;

    for (let right = 0; right < s.length; right++) {
        if (s[right] === '1') {
            ones++;
        }

        while (ones === k) {
            while (s[left] === '0') {
                left++;
            }

            let window = s.slice(left, right + 1);
            if (
                ans === "" ||
                window.length < ans.length ||
                (window.length === ans.length && window < ans)
            ) {
                ans = window;
            }

            ones--;
            left++;
        }

    }

    return ans;
};

File: 2904-shortest-and-lexicographically-smallest-beautiful-string/2904-shortest-and-lexicographically-smallest-beautiful-string.py
class Solution:
    def shortestBeautifulSubstring(self, s: str, k: int) -> str:
        ans = ""
        ones = 0
        left = 0

        for right in range(len(s)):
            if s[right] == "1":
                ones += 1

            while ones == k:
                while s[left] == "0":
                    left += 1

                window = s[left : right + 1]

                if (
                    ans == ""
                    or len(window) < len(ans)
                    or (len(window) == len(ans) and window < ans)
                ):
                    ans = window

                left += 1
                ones -= 1

        return ans
